_drafts/c-struct-padding-initialization.md

```diff
@@ -0,0 +1,460 @@
+
+If the structure is only ever accessed on a per-member basis, the padding
+probably won't cause problems:
+
+```c
+struct foo {
+  uint32_t i;
+  uint8_t b;
+};
+
+// Check 2 foos for equality
+bool foo_are_equal(struct foo a, struct foo b) {
+  return (a.i == b.i) && (a.b == b.b);
+}
+```
+
+If you are serializing the structure outside of your application (eg, over a
+UART, or writing it to non-volatile storage), but the data is _only_ ever read
+by your application, there's a good chance the padding bits won't matter.
+
+If on the other hand the structure needs to be parsed by another piece of
+software, it might be preferrable to pack the structure for easy to understand
+layout, at the expense of computation.
+
+## Structure (Zero) Initialization
+
+Given the above, it seems convenient to zero-intialize structures before using
+them. In C99+, the following patterns are common for structure initialization:
+
+```c
+struct foo {
+  uint32_t i;
+  uint8_t b;
+};
+
+// Initialize members of 'a' to specific values. Members not specifically
+// initialized will be initialized per the 'static storage duration'
+// initialization rules (eg pointers go to NULL, integers go to 0, floats go to
+// 0.0, etc)
+struct foo a = {
+  .i = 1,
+  // .b will be set to 0
+};
+
+// Initialize 'b' to all zeros. This is a common idiom that really just
+// specifies a '0' constant as the initial value for the first member of the
+// structure, then relies on the above rule to initialize the rest of the
+// structure.
+struct foo b = { 0 };
+```
+
+This looks great! However, it's not obvious (from looking at those snippets)
+what the value loaded into the padding region will be.
+
+The unfortunate answer is: _it depends_
+
+The C11 standard, chapter §6.2.6.1/6 says this:
+
+> When a value is stored in an object of structure or union type, including in a
+> member object, the bytes of the object representation that correspond to any
+> padding bytes take unspecified values.
+
+See also <https://stackoverflow.com/a/37642061>
+
+Objects with '_static_ storage duration' (`static` keyword, or external linkage
+(defined at the outermost scope in a compilation unit)), **padding bits will be
+initialized to 0**!
+
+Objects with '_automatic_ storage duration' (locally-scoped objects) have
+**undefined behavior** when it comes to padding bit initialization!
+
+This means that there is no constraint on what values are set to those bits when
+the object is initialized.
+
+## The Current State
+
+Let's consider the following 4 initialization strategies for this structure:
+
+```c
+struct foo {
+  uint32_t i;
+  uint8_t b;
+};
+```
+
+1. `memset` to zeros:
+
+   ```c
+   struct foo a;
+   memset(&a, 0, sizeof(a));
+   ```
+
+2. individually set all members to 0:
+
+   ```c
+   struct foo a = {
+     .i = 0,
+     .b = 0,
+   };
+   ```
+
+3. use `{ 0 }` zero-initializer
+
+   ```c
+   struct foo a = { 0 };
+   ```
+
+4. use `{}` GCC extension zero-initializer
```

Comment:
heh I was going to say list the doc to this extension but [apparently there is not one](https://gcc.gnu.org/legacy-ml/gcc/2019-07/msg00065.html). May be worth noting this syntax is valid C++ (just not C)

Reply:
sweet, added the reference! i can't even find documentation on this feature in the gcc sources (predates git!) :open_mouth: very secret corner of the language